Feat(LP-35165): Nxf version depending on environment + rich click text

danielboloc opened the pull request:
# Overview

- Adds `click` rich text
- Updates and restricts Nextflow version depending on the execution platform
  - 'aws' + '22.10.8'     ==> '22.10.8'
  - 'aws' + '24.04.4'    ==> '24.04.4'
  - 'aws' + 'latest'        ==> '24.04.4'
  - 'azure' + '22.10.8'  ==> '22.11.1-edge'
  - 'azure' + '24.04.4' ==> '22.11.1-edge'
  - 'azure' + 'latest'     ==> '22.11.1-edge'
  - 'hpc' + '22.10.8'     ==> '22.10.8'
  - 'hpc' + '24.04.4'    ==> '22.10.8'
  - 'hpc' + 'latest'        ==> '22.10.8'

Closes PR #182 as it includes the changes here.
Closes issue #31 .

# JIRA

https://lifebit.atlassian.net/browse/LP-35165

# Tests

Docker container created and pushed to Quay `quay.io/lifebitaiorg/cloudos-cli:v2.19.0`

## AWS & `22.11.1-edge` ==> `22.10.8`

WORKSPACE_ID="5c6d3e9bd954e800b23f8c62"
MY_API_KEY="****"
CLOUDOS="https://cloudos.lifebit.ai"
WORKFLOW_NAME="rnatoy"
PROJECT_NAME="DB-rna"
NEXTFLOW_PROFILE="test"

```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-version '22.11.1-edge' \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable

CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] No job queue was specified, using the CloudOS default queue: New_standard_cost_saving.
[Message] For execution platform 'aws', the workflow will use the default '22.10.8' version on CloudOS.

Executing run...
        Nextflow version: 22.10.8
        Job successfully launched to CloudOS, please check the following link: https://cloudos.lifebit.ai/app/jobs/67e5652f833a7c1747eb20b8
        Your assigned job id is: 67e5652f833a7c1747eb20b8

        Your current job status is: initializing
        To further check your job status you can either go to https://cloudos.lifebit.ai/app/jobs/67e5652f833a7c1747eb20b8 or use the following command:
        cloudos job status \
                --apikey $MY_API_KEY \
                --cloudos-url https://cloudos.lifebit.ai \
                --job-id 67e5652f833a7c1747eb20b8
```
<img width="1346" alt="image" src="https://github.com/user-attachments/assets/6bc428dd-e75a-4268-b096-d004f35c0d96" />

## AWS & `latest` ==> `24.04.4`

```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-version 'latest' \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable

CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] No job queue was specified, using the CloudOS default queue: New_standard_cost_saving.
[Message] You have specified Nextflow version 'latest' for execution platform 'aws'. The workflow will use the latest version available on CloudOS: 24.04.4.
[Warning] You have specified Nextflow version 24.04.4. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
        Nextflow version: 24.04.4
        Job successfully launched to CloudOS, please check the following link: https://cloudos.lifebit.ai/app/jobs/67e565dc833a7c1747eb2b76
        Your assigned job id is: 67e565dc833a7c1747eb2b76

        Your current job status is: initializing
        To further check your job status you can either go to https://cloudos.lifebit.ai/app/jobs/67e565dc833a7c1747eb2b76 or use the following command:
        cloudos job status \
                --apikey $MY_API_KEY \
                --cloudos-url https://cloudos.lifebit.ai \
                --job-id 67e565dc833a7c1747eb2b76
```
> [!NOTE]
> It seems that the `rnatoy` does not work with Nextflow `24.04.4`, but the job is submitted correctly
<img width="1340" alt="image" src="https://github.com/user-attachments/assets/6ca11bb3-8302-43f9-98dd-bd3b598bc89c" />


## Azure & `22.10.8` ==> `22.11.1-edge`

WORKSPACE_ID="6480f3db916489d248956a5f"
MY_API_KEY="****"
CLOUDOS="https://dev.sdlc.lifebit.ai"
WORKFLOW_NAME="cufflink"
PROJECT_NAME="testdata_azure"
NEXTFLOW_PROFILE="test"

```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-version '22.10.8' \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable \
    --execution-platform azure

CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] For execution platform 'azure', the workflow will use the '22.11.1-edge' version on CloudOS.
[Warning] You have specified Nextflow version 22.11.1-edge. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
        Nextflow version: 22.11.1-edge
        Job successfully launched to CloudOS, please check the following link: https://dev.sdlc.lifebit.ai/app/jobs/67e5665c8fc6e9f297a387fe
        Your assigned job id is: 67e5665c8fc6e9f297a387fe

        Your current job status is: initializing
        To further check your job status you can either go to https://dev.sdlc.lifebit.ai/app/jobs/67e5665c8fc6e9f297a387fe or use the following command:
        cloudos job status \
                --apikey $MY_API_KEY \
                --cloudos-url https://dev.sdlc.lifebit.ai \
                --job-id 67e5665c8fc6e9f297a387fe
```
<img width="1362" alt="image" src="https://github.com/user-attachments/assets/34725bbf-29db-4f53-b188-0604e89b6497" />


## Azure & no version added

```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable \
    --execution-platform azure

CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] For execution platform 'azure', the workflow will use the '22.11.1-edge' version on CloudOS.
[Warning] You have specified Nextflow version 22.11.1-edge. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
        Nextflow version: 22.11.1-edge
        Job successfully launched to CloudOS, please check the following link: https://dev.sdlc.lifebit.ai/app/jobs/67e56680606a74eb08ff1393
        Your assigned job id is: 67e56680606a74eb08ff1393

        Your current job status is: initializing
        To further check your job status you can either go to https://dev.sdlc.lifebit.ai/app/jobs/67e56680606a74eb08ff1393 or use the following command:
        cloudos job status \
                --apikey $MY_API_KEY \
                --cloudos-url https://dev.sdlc.lifebit.ai \
                --job-id 67e56680606a74eb08ff1393
```
<img width="1348" alt="image" src="https://github.com/user-attachments/assets/e39c4649-0f79-4299-be9f-d77d0912bdeb" />


## Azure & `latest` ==> `22.11.1-edge`

```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --nextflow-version 'latest' \
    --resumable \
    --execution-platform azure

CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] You have specified Nextflow version 'latest' for execution platform 'azure'. The workflow will use the latest version available on CloudOS: 22.11.1-edge.
[Warning] You have specified Nextflow version 22.11.1-edge. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
        Nextflow version: 22.11.1-edge
        Job successfully launched to CloudOS, please check the following link: https://dev.sdlc.lifebit.ai/app/jobs/67e566d9606a74eb08ff19a9
        Your assigned job id is: 67e566d9606a74eb08ff19a9

        Your current job status is: initializing
        To further check your job status you can either go to https://dev.sdlc.lifebit.ai/app/jobs/67e566d9606a74eb08ff19a9 or use the following command:
        cloudos job status \
                --apikey $MY_API_KEY \
                --cloudos-url https://dev.sdlc.lifebit.ai \
                --job-id 67e566d9606a74eb08ff19a9
```
<img width="1327" alt="image" src="https://github.com/user-attachments/assets/536de41a-9375-46f3-9aed-c853ac396064" />


## adding a wrong Nextflow version
<img width="1252" alt="image" src="https://github.com/user-attachments/assets/c26298af-484f-4786-b1ad-074d356206da" />

## `rich_click`

### General

<img width="1429" alt="image" src="https://github.com/user-attachments/assets/0cf80700-efe9-4c89-a333-fc3f99c10036" />

### Queue
<img width="1428" alt="image" src="https://github.com/user-attachments/assets/6d747f06-c40e-428c-8922-c7ce7c330091" />
<img width="1435" alt="image" src="https://github.com/user-attachments/assets/dacd5e10-49dd-46a9-bd19-46aa5c305262" />

### Job

<img width="1430" alt="image" src="https://github.com/user-attachments/assets/0642e523-7b4b-46c1-8d29-5ad6413476a7" />

### Project
<img width="1431" alt="image" src="https://github.com/user-attachments/assets/e9a82ffe-7db8-4f0b-8aaf-433efcfa0cbf" />


### Not found option
<img width="1430" alt="image" src="https://github.com/user-attachments/assets/22601b1b-57f3-4d38-8bc0-3a87195051cb" />

dapineyro commented on cloudos_cli/__main__.py:
```diff
@@ -410,9 +410,26 @@ def run(apikey,
+        print(f'[Message] You have specified Nextflow version \'latest\' for execution platform \'{execution_platform}\'. The workflow will use the ' +
               f'latest version available on CloudOS: {nextflow_version}.')
+    if execution_platform == 'aws':
+        if nextflow_version not in ['24.04.4', '22.10.8']:
```
```suggestion
        if nextflow_version not in AWS_NEXTFLOW_VERSIONS:
```

Could we add three new global vars at the beginning of the script:
```
AWS_NEXTFLOW_VERSIONS = ['22.10.8', '24.04.4']
AZURE_NEXTFLOW_VERSIONS = ['22.11.1-edge']
HPC_NEXTFLOW_VERSIONS = ['22.10.8']
```
Doing so we can easily update. For the same reason, it would also be nice to have another set of global vars for the latest version of each execution_platform:
```
AWS_NEXTFLOW_LATEST = '24.04.4'
AZURE_NEXTFLOW_LATEST = '22.11.1-edge4'
HPC_NEXTFLOW_LATEST = '22.10.8'
```

danielboloc commented:
@dapineyro I have updated the code with your suggestion and I have updated the Docker container

`docker run -it --rm quay.io/lifebitaiorg/cloudos-cli:v2.19.0 bash`

# AWS 

## AWS & `22.11.1-edge` ==> `22.10.8`
```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-version '22.11.1-edge' \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable
CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] No job queue was specified, using the CloudOS default queue: New_standard_cost_saving.
[Message] For execution platform 'aws', the workflow will use the default '22.10.8' version on CloudOS.

Executing run...
	Nextflow version: 22.10.8
	Job successfully launched to CloudOS, please check the following link: https://cloudos.lifebit.ai/app/jobs/67e581f5519a9f3252a535a4
	Your assigned job id is: 67e581f5519a9f3252a535a4

	Your current job status is: initializing
	To further check your job status you can either go to https://cloudos.lifebit.ai/app/jobs/67e581f5519a9f3252a535a4 or use the following command:
	cloudos job status \
		--apikey $MY_API_KEY \
		--cloudos-url https://cloudos.lifebit.ai \
		--job-id 67e581f5519a9f3252a535a4
```

## AWS & `latest` ==> `24.04.4`
```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-version 'latest' \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable
CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] No job queue was specified, using the CloudOS default queue: New_standard_cost_saving.
[Message] You have specified Nextflow version 'latest' for execution platform 'aws'. The workflow will use the latest version available on CloudOS: 24.04.4.
[Warning] You have specified Nextflow version 24.04.4. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
	Nextflow version: 24.04.4
	Job successfully launched to CloudOS, please check the following link: https://cloudos.lifebit.ai/app/jobs/67e5824d519a9f3252a54d57
	Your assigned job id is: 67e5824d519a9f3252a54d57

	Your current job status is: initializing
	To further check your job status you can either go to https://cloudos.lifebit.ai/app/jobs/67e5824d519a9f3252a54d57 or use the following command:
	cloudos job status \
		--apikey $MY_API_KEY \
		--cloudos-url https://cloudos.lifebit.ai \
		--job-id 67e5824d519a9f3252a54d57
```

# Azure

## Azure & `22.10.8` ==> `22.11.1-edge`
```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-version '22.10.8' \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable \
    --execution-platform azure
CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] For execution platform 'azure', the workflow will use the '22.11.1-edge' version on CloudOS.
[Warning] You have specified Nextflow version 22.11.1-edge. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
	Nextflow version: 22.11.1-edge
	Job successfully launched to CloudOS, please check the following link: https://dev.sdlc.lifebit.ai/app/jobs/67e5828022842aea5f3d9933
	Your assigned job id is: 67e5828022842aea5f3d9933

	Your current job status is: initializing
	To further check your job status you can either go to https://dev.sdlc.lifebit.ai/app/jobs/67e5828022842aea5f3d9933 or use the following command:
	cloudos job status \
		--apikey $MY_API_KEY \
		--cloudos-url https://dev.sdlc.lifebit.ai \
		--job-id 67e5828022842aea5f3d9933
```

## Azure & no version added
```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --resumable \
    --execution-platform azure
CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] For execution platform 'azure', the workflow will use the '22.11.1-edge' version on CloudOS.
[Warning] You have specified Nextflow version 22.11.1-edge. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
	Nextflow version: 22.11.1-edge
	Job successfully launched to CloudOS, please check the following link: https://dev.sdlc.lifebit.ai/app/jobs/67e582a822842aea5f3da1cd
	Your assigned job id is: 67e582a822842aea5f3da1cd

	Your current job status is: initializing
	To further check your job status you can either go to https://dev.sdlc.lifebit.ai/app/jobs/67e582a822842aea5f3da1cd or use the following command:
	cloudos job status \
		--apikey $MY_API_KEY \
		--cloudos-url https://dev.sdlc.lifebit.ai \
		--job-id 67e582a822842aea5f3da1cd
```

## Azure & `latest` ==> `22.11.1-edge`
```console
cloudos job run \
    --cloudos-url $CLOUDOS \
    --apikey $MY_API_KEY \
    --workspace-id $WORKSPACE_ID \
    --project-name "$PROJECT_NAME" \
    --workflow-name $WORKFLOW_NAME \
    --nextflow-profile $NEXTFLOW_PROFILE \
    --nextflow-version 'latest' \
    --resumable \
    --execution-platform azure
CloudOS python package: a package for interacting with CloudOS.

Version: 2.19.0

CloudOS job functionality: run and check jobs in CloudOS.

[Message] You have specified Nextflow version 'latest' for execution platform 'azure'. The workflow will use the latest version available on CloudOS: 22.11.1-edge.
[Warning] You have specified Nextflow version 22.11.1-edge. This version requires the pipeline to be written in DSL2 and does not support DSL1.

Executing run...
	Nextflow version: 22.11.1-edge
	Job successfully launched to CloudOS, please check the following link: https://dev.sdlc.lifebit.ai/app/jobs/67e582c922842aea5f3da8ff
	Your assigned job id is: 67e582c922842aea5f3da8ff

	Your current job status is: initializing
	To further check your job status you can either go to https://dev.sdlc.lifebit.ai/app/jobs/67e582c922842aea5f3da8ff or use the following command:
	cloudos job status \
		--apikey $MY_API_KEY \
		--cloudos-url https://dev.sdlc.lifebit.ai \
		--job-id 67e582c922842aea5f3da8ff
```

dapineyro reviewed (approved):
Thanks @danielboloc !
